src/types/widget.ts

```diff
@@ -110,102 +111,167 @@ export type BuiltinWidgetTypes =
   | 'ais.toggleRefinement'
   | 'ais.voiceSearch';
 
-/**
- * Widgets are the building blocks of InstantSearch.js. Any valid widget must
- * have at least a `render` or a `init` function.
- */
-export type Widget<
-  TWidgetOptions extends { renderState: unknown } = { renderState: unknown }
-> = {
-  /**
-   * Identifier for official connectors and widgets
-   */
-  $$type?: BuiltinTypes;
+export type WidgetParams = {
+  widgetParams: NonNullable<unknown>;
+};
 
+export type WidgetDescription = {
+  $$type: string;
+  $$widgetType?: string;
+  renderState?: Record<string, unknown>;
+  indexRenderState?: Record<string, unknown>;
+  indexUiState?: Record<string, unknown>;
+};
+
+type RequiredWidgetLifeCycle<TWidgetDescription extends WidgetDescription> = {
   /**
-   * Identifier for official widgets
+   * Identifier for connectors and widgets.
    */
-  $$widgetType?: BuiltinWidgetTypes;
+  $$type: TWidgetDescription['$$type'];
 
   /**
-   * Called once before the first search
+   * Called once before the first search.
    */
-  init?(options: InitOptions): void;
+  init?: (options: InitOptions) => void;
   /**
-   * Called after each search response has been received
+   * Called after each search response has been received.
    */
-  render?(options: RenderOptions): void;
+  render?: (options: RenderOptions) => void;
```

Comment:
A widget must have at least one of [init, render]. What if we extract this part and use `AtLeastOne` type util?

Reply:
I thought about that, but in practice that would make the type more complicated for few gain, since $$type is required now, and it doesn't really make the usage simpler (will still need to use `init!()`, since you can't know which one of the two is actually implemented). Instead I think we should in a major version make render required both here and in addWidgets;

Reply:
Oh yeah, it makes sense!